File: win-office-com-automation/DocReaderCli/Readers/PowerPointReader.cs
using System.Text;
using NetOffice.PowerPointApi;
using NetOffice.OfficeApi.Enums;

namespace DocReaderCli.Readers;

public static class PowerPointReader
{
    private const int DrmPollIntervalMs = 500;
    private const int DrmTimeoutMs = 15_000;

    public static string Read(string filePath)
    {
        using var watchdog = new ProcessWatchdog("POWERPNT");
        Application? app = null;
        Presentation? pres = null;

        try
        {
            app = new Application();
            // PowerPoint doesn't support Visible=false at app level in all versions,
            // so we open the file with window hidden.
            watchdog.DetectNewProcess();

            // Presentations.Open positional: FileName, ReadOnly, Untitled, WithWindow
            pres = app.Presentations.Open(
                filePath,
                NetOffice.OfficeApi.Enums.MsoTriState.msoTrue,   // ReadOnly
                NetOffice.OfficeApi.Enums.MsoTriState.msoFalse,  // Untitled
                NetOffice.OfficeApi.Enums.MsoTriState.msoFalse   // WithWindow
            );

            WaitForDrmDecryption(pres, watchdog.TimeoutMs);

            var sb = new StringBuilder();

            int slideNum = 0;
            foreach (Slide slide in pres.Slides)
            {
                slideNum++;
                sb.AppendLine($"## Slide {slideNum}");
                sb.AppendLine();

                try
                {
                    ExtractSlide(slide, sb);
                }
                catch (Exception ex)
                {
                    Console.Error.WriteLine($"[PPTReader] Error on slide {slideNum}: {ex.Message}");
                }

                sb.AppendLine();
            }

            return sb.ToString();
        }
        finally
        {
            try { pres?.Close(); } catch { }
            try { app?.Quit(); } catch { }
            try { app?.Dispose(); } catch { }
            watchdog.KillIfRunning();
        }
    }

    private static void WaitForDrmDecryption(Presentation pres, int timeoutMs)
    {
        var sw = System.Diagnostics.Stopwatch.StartNew();
        while (sw.ElapsedMilliseconds < timeoutMs)
        {
            try
            {
                if (pres.Slides.Count > 0)
                {
                    // Try to access first slide's shapes
                    var firstSlide = pres.Slides[1];
                    if (firstSlide.Shapes.Count >= 0)
                        return;
                }
            }
            catch { }

            Thread.Sleep(DrmPollIntervalMs);
        }

        // Allow empty presentations
        try
        {
            if (pres.Slides.Count == 0)
                return;
        }
        catch { }

        throw new TimeoutException(
            $"DRM decryption timed out after {timeoutMs / 1000}s.");
    }

    private static void ExtractSlide(Slide slide, StringBuilder sb)
    {
        // Extract title if present
        try
        {
            if (slide.Shapes.HasTitle == NetOffice.OfficeApi.Enums.MsoTriState.msoTrue)
            {
                var titleText = slide.Shapes.Title?.TextFrame?.TextRange?.Text;
                if (!string.IsNullOrWhiteSpace(titleText))
                {
                    sb.AppendLine($"### {titleText.Trim()}");
                    sb.AppendLine();
                }
            }
        }
        catch { }

        // Extract all text shapes
        foreach (Shape shape in slide.Shapes)
        {
            try
            {
                if (shape.HasTextFrame == MsoTriState.msoTrue)
                {
                    var text = shape.TextFrame?.TextRange?.Text;
                    if (!string.IsNullOrWhiteSpace(text))
                    {
                        // Skip if this is the title (already emitted)
                        try
                        {
                            if (slide.Shapes.HasTitle == MsoTriState.msoTrue &&
                                shape.Name == slide.Shapes.Title?.Name)
                                continue;
                        }
                        catch { }

                        sb.AppendLine(text.Trim());
                        sb.AppendLine();
                    }
                }

                // Extract text from tables in shapes
                if (shape.HasTable == MsoTriState.msoTrue)
                {
                    ExtractTable(shape.Table, sb);
                }
            }
            catch { }
        }

        // Extract notes
        try
        {
            var notes = slide.NotesPage?.Shapes;
            if (notes != null)
            {
                foreach (Shape noteShape in notes)
                {
                    try
                    {
                        if (noteShape.HasTextFrame == MsoTriState.msoTrue)
                        {
                            var noteText = noteShape.TextFrame?.TextRange?.Text;
                            if (!string.IsNullOrWhiteSpace(noteText) &&
                                noteText.Trim().Length > 1) // skip slide number placeholder
                            {
                                sb.AppendLine($"> **Note:** {noteText.Trim()}");
                                sb.AppendLine();
                            }
                        }
                    }
                    catch { }
                }
            }
        }
        catch { }
    }

    private static void ExtractTable(NetOffice.PowerPointApi.Table table, StringBuilder sb)
    {
        try
        {
            int rows = table.Rows.Count;
            int cols = table.Columns.Count;

            sb.AppendLine();
            for (int r = 1; r <= rows; r++)
            {
                sb.Append("|");
                for (int c = 1; c <= cols; c++)
                {
                    string cellText = "";
                    try
                    {
                        cellText = table.Cell(r, c).Shape.TextFrame.TextRange.Text ?? "";
                        cellText = cellText.Replace("|", "\\|").Replace("\n", " ").Replace("\r", "");
                    }
                    catch { }
                    sb.Append($" {cellText} |");
                }
                sb.AppendLine();

                if (r == 1)
                {
                    sb.Append("|");
                    for (int c = 0; c < cols; c++)
                        sb.Append(" --- |");
                    sb.AppendLine();
                }
            }
            sb.AppendLine();
        }
        catch { }
    }
}

File: win-office-read-py/DocReaderCliPy/docreader_cli/readers/powerpoint_reader.py
from __future__ import annotations

import os
from typing import Any

from ..markdown import emit_markdown_table
from ..office_app import log, wait_for_active_app, wait_for_matching_document, wait_until_ready
from ..process_watchdog import ProcessWatchdog


DRM_TIMEOUT_SECONDS = 15
MSO_TRUE = -1


def read_powerpoint(file_path: str) -> str:
    with ProcessWatchdog("POWERPNT") as watchdog:
        app = None
        presentation = None
        try:
            log(f"[PPTReader] Opening presentation via shell: {file_path}")
            os.startfile(file_path)
            watchdog.detect_new_process()

            app = wait_for_active_app("PowerPoint.Application", watchdog.timeout_ms / 1000)
            presentation = wait_for_matching_document(
                app.Presentations,
                file_path,
                watchdog.timeout_ms / 1000,
                "Presentation",
            )

            wait_until_ready(
                lambda: _is_powerpoint_ready(presentation),
                DRM_TIMEOUT_SECONDS,
                "DRM decryption timed out after 15s.",
            )
            return _extract_presentation(presentation)
        finally:
            if presentation is not None:
                try:
                    presentation.Close()
                except Exception:
                    pass


def _is_powerpoint_ready(presentation: Any) -> bool:
    if int(presentation.Slides.Count) == 0:
        return True
    first_slide = presentation.Slides.Item(1)
    _ = first_slide.Shapes.Count
    return True


def _extract_presentation(presentation: Any) -> str:
    parts: list[str] = []
    slide_count = int(presentation.Slides.Count)
    for index in range(1, slide_count + 1):
        try:
            slide = presentation.Slides.Item(index)
            parts.append(_extract_slide(slide, index))
        except Exception as exc:
            log(f"[PPTReader] Error on slide {index}: {exc}")
    return "".join(parts)


def _extract_slide(slide: Any, slide_number: int) -> str:
    parts = [f"## Slide {slide_number}\n\n"]
    title_name = None

    try:
        if slide.Shapes.HasTitle == MSO_TRUE:
            title_shape = slide.Shapes.Title
            title_name = title_shape.Name
            title_text = title_shape.TextFrame.TextRange.Text
            if title_text and str(title_text).strip():
                parts.append(f"### {str(title_text).strip()}\n\n")
    except Exception:
        pass

    shape_count = int(slide.Shapes.Count)
    for index in range(1, shape_count + 1):
        try:
            shape = slide.Shapes.Item(index)
            if title_name and getattr(shape, "Name", None) == title_name:
                continue

            if getattr(shape, "HasTextFrame", 0) == MSO_TRUE:
                text = shape.TextFrame.TextRange.Text
                if text and str(text).strip():
                    parts.append(f"{str(text).strip()}\n\n")

            if getattr(shape, "HasTable", 0) == MSO_TRUE:
                parts.append(_extract_table(shape.Table))
        except Exception:
            continue

    try:
        note_shapes = slide.NotesPage.Shapes
        note_count = int(note_shapes.Count)
        for index in range(1, note_count + 1):
            try:
                note_shape = note_shapes.Item(index)
                if getattr(note_shape, "HasTextFrame", 0) == MSO_TRUE:
                    note_text = note_shape.TextFrame.TextRange.Text
                    if note_text and len(str(note_text).strip()) > 1:
                        parts.append(f"> **Note:** {str(note_text).strip()}\n\n")
            except Exception:
                continue
    except Exception:
        pass

    return "".join(parts)


def _extract_table(table: Any) -> str:
    rows = int(table.Rows.Count)
    cols = int(table.Columns.Count)
    matrix: list[list[str]] = []
    for row in range(1, rows + 1):
        values: list[str] = []
        for col in range(1, cols + 1):
            try:
                values.append(str(table.Cell(row, col).Shape.TextFrame.TextRange.Text or ""))
            except Exception:
                values.append("")
        matrix.append(values)
    return emit_markdown_table(matrix)
